src/server/transaction.cc

```diff
@@ -796,8 +779,8 @@ void Transaction::UnlockMulti() {
 
   use_count_.fetch_add(shard_data_.size(), std::memory_order_relaxed);
   for (ShardId i = 0; i < shard_data_.size(); ++i) {
-    shard_set->Add(i, [this, sharded_keys, shard_journals_cnt]() {
-      this->UnlockMultiShardCb(*sharded_keys, EngineShard::tlocal(), shard_journals_cnt);
+    shard_set->Add(i, [this, sharded_keys, i, shard_journals_cnt]() {
+      this->UnlockMultiShardCb((*sharded_keys)[i], EngineShard::tlocal(), shard_journals_cnt);
       intrusive_ptr_release(this);
     });
   }
```

Comment:
just a reminder: one potential optimization is skipping this callback if the list is empty. I assume most tx'es don't have keys on all shards